src/validators/keystores/hashi_vault.py

```diff
@@ -109,3 +120,133 @@ async def _load_hashi_vault_keys(session: ClientSession, secret_url: str) -> Key
+                *[
+                    self._load_bundled_hashi_vault_keys(
+                        session=session,
+                        secret_url=self.config.secret_url(key_path),
+                    )
+                    for key_path in input_chunk
+                ]
+            )
+        self.merge_keys_responses(keys_responses, merged_keys)
+
+
+@dataclass
+class PrefixedKeysLoadedCallback:
+    """Future done callback, that appends keys to shared dict."""
+
+    prefix: str
+    mapping: dict[str, list[str]]
+
+    def __call__(self, keys_future_resolved: asyncio.Future) -> None:
+        self.mapping.update({self.prefix: keys_future_resolved.result()})
+
+
+class HashiVaultPrefixedKeysLoader(HashiVaultKeysLoader):
+    @staticmethod
+    async def _find_prefixed_hashi_vault_keys(session: ClientSession, prefix_url: str) -> list[str]:
+        """
+        Discover public keys under prefix in hashi vault K/V secret engine
+
+        All public keys must be a final chunk of the secret path without 0x prefix,
+        all secret keys are stored under these paths with arbitrary secret dictionary
+        key, and secret value with or without 0x prefix.
+        """
+        logger.info('Will discover validator keys in %s', prefix_url)
+        response = await session.request(method='LIST', url=prefix_url)
+        response.raise_for_status()
+        key_paths = await response.json()
+        if 'data' not in key_paths:
+            logger.error('Failed to discover keys in hashi vault')
+            for error in key_paths.get('errors', []):
+                logger.error('hashi vault error: %s', error)
+            raise RuntimeError('Can not discover validator public keys from hashi vault')
+        return key_paths['data']['keys']
+
+    @staticmethod
+    async def _load_prefixed_hashi_vault_key(session: ClientSession, secret_url: str) -> Keys:
+        logger.info('Will load keys from %s', secret_url)
+        response = await session.get(url=secret_url)
+        response.raise_for_status()
+        key_data = await response.json()
+        if 'data' not in key_data:
+            logger.error('Failed to retrieve keys from hashi vault')
+            for error in key_data.get('errors', []):
+                logger.error('hashi vault error: %s', error)
+            raise RuntimeError('Can not retrieve validator signing keys from hashi vault')
+        # Last chunk of URL is a public key
+        pk = secret_url.strip('/').split('/')[-1].strip('0x')
+        if len(key_data['data']['data']) > 1:
+            raise RuntimeError(
+                f'Invalid multi-value secret at path {secret_url}, '
+                'should only contain single value',
+            )
+        sk = list(key_data['data']['data'].values())[0]
+        sk_bytes = Web3.to_bytes(hexstr=sk)
+        return Keys({add_0x_prefix(HexStr(pk)): BLSPrivkey(sk_bytes)})
+
+    async def process_keys_chunk(self, input_chunk: list[str], merged_keys: Keys) -> None:
```

Comment:
Usually "public" methods are put at the top of class, and "internal" methods (starting from underscore) are below.

Reply:
Example reference: https://pypi.org/project/flake8-newspaper-style/